dom.bs

```diff
@@ -2192,6 +2192,10 @@ steps:</p>
  <li><p>If the <i>open flag</i> is set and <var>shadow</var>'s <a for=ShadowRoot>mode</a> is
  <em>not</em> "<code>open</code>", then return null.</p></li>
 
+ <li><p>If <var>shadow</var>'s <a for=ShadowRoot>slot assignment</a> is "<code>manual</code>",
+ then return the <a>slot</a> in <var>shadow</var>'s <a for=tree>descendants</a> whose <a>manually assigned nodes</a>
```

Comment:
This probably needs to find the first slot because there could be multiple slots referring to the same node (using element reflection style weak reference would mean that we also want to allow re-ordering of slots without losing their assigned nodes.

Reply:
Hmm, if I'm thinking about this correctly, there should only be one slot referring to a given node, because the [assign](https://github.com/whatwg/html/pull/5483/files#diff-41cf6794ba4200b839c53531555f0f3998df4cbb01a4d5cb0b94e3ca5e23947dR59369) method loops through all slots within the shadow root and removes the given nodes from all other slots. It also de-duplicates assigned nodes within a single slot. So I think this algorithm should always return a unique slot. Correct me if I'm wrong.

Separately, I do think we need to add a step to [remove a node](https://dom.spec.whatwg.org/#concept-node-remove) that clears out all manually assigned nodes from `<slot>` elements when they get removed from the tree. Without that, there is the definite possibility of the multiple slot problem you bring up, because the `assign()` steps won't be able to find all slots that might point to a node.

Reply:
I mean... that would defeat the whole point of using weak reference defined for element reflection. The whole point of using that technology is so that slot can be removed & inserted into another place without losing its assigned nodes.

Reply:
Hmm, I thought the purpose was allowing the movement of **light dom** children of the host, not the `<slot>` elements themselves. Is that an important use case?

The (potentially huge?) downside of allowing movement of both light dom children **and** slot elements is that the `manually assigned nodes` vector can start to contain duplicates all over the place. Since that vector is not developer-visible, it might get very confusing, if the order of the `<slot>` elements starts to dictate which manually-assigned node gets assigned to which slot. Consider:

```html
<div id=host>
  <template shadowroot=open shadowrootmanualslotting>
    <slot id=slot1></slot>
    <slot id=slot2></slot>
  </template>
  <div id=node1></div>
  <div id=node2></div>
</div>

<script>
slot1.assign([node1, node2]);
slot1.remove();
slot2.assign([node1, node2]);
// Here, both node1/node2 will be assigned to `slot2`.
host.shadowRoot.appendChild(slot1);
// Still, both node1/node2 will be assigned to `slot2`, because `slot1` comes last.
host.shadowRoot.appendChild(slot2);
// Now, both node1/node2 will be assigned to `slot1`.
</script>
```

It seems more tractable (from a developer's point of view) to have all slots within a shadow root contain unique sets of manually assigned nodes, no? Otherwise, they'll always have to call `.assign()` on any slots they append to the shadow root, to make sure the (invisible) manually assigned nodes are still what they think.

Reply:
The simplest solution to that problem is that whenever a node is assigned to a new slot, then it gets removed from all other slots.

Reply:
It is problematic to require the list of nodes that can be assigned to a given slot to be in the shadow tree of their parent node as that would force the slot elements to be in the shadow tree before any nodes can be assigned. That is the kind of construction relationship problem we discussed during AOM / element reflection and concluded that it is a show stopper issue for that use case. I don't see why it would be any different here.

Reply:
> The simplest solution to that problem is that whenever a node is assigned to a new slot, then it gets removed from all other slots.

Hmm ok, this seems like a reasonable solution. It does require another `weak_ptr` from `Node` back to its assigned slot. I suppose this isn't too bad. If you're ok with that solution, I'll see if I can add it to the spec. It definitely simplifies things to have only one slot per node.

Reply:
Right. Only caveat is that a given node can be simultaneously assigned to a slot of a shadow tree which uses automatic "name" based assignment and to a slot which is not in that shadow tree. In that case, the manual assignment should work but the automatic name based assignment should prevail since the manual assignment is not in effect as the assigned node is not a slottable of the host of a shadow root of the slot to which it was manually assigned. But it is true that at any given point, each node can only be assigned to at most one slot manually.

There is a bit of oddity here in that each shadow tree determines which mode of assignment we're using and slot doesn't despite of the fact manual assignment API appears on slot element. So in theory, it's possible for someone to assign a node to a slot and then insert that into a shadow tree which uses automatic name based assignment. I guess we'd ignore the manual assignment in that case.

Reply:
Ok great. And that's a great point about a slot with manually assigned nodes that is living in a shadow host in "named" mode. I agree with the behavior you proposed, and I'll try to get it written up.

There are a number of potentially confusing things that the developer can do here, and I'm wondering which of them (if any) should get helpful console warnings. I feel like there will be a tough balance to strike between log spam and a lack of understanding of what's going on.

Reply:
That is a fair point. You might want to sync up with @alice about this.

Reply:
Will do. I reached out to her privately to see what she thinks.